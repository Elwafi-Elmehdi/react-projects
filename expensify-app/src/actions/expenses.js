import { v4 as uuidv4 } from "uuid";

// Actions for Expensify

// ADD_EXPENSE
export const addExpense = ({
	title = "no title",
	note = "",
	amount = 0.0,
	createdAt = 0,
} = {}) => ({
	type: "ADD_EXPENSE",
	expense: {
		id: uuidv4(),
		title,
		note,
		amount,
		createdAt,
	},
});

// REMOVE_EXPENSE
export const removeExpense = ({ id } = {}) => ({
	type: "REMOVE_EXPENSE",
	id,
});
// EDIT_EXPENSE
export const editExpense = (id, updates) => ({
	type: "EDIT_EXPENSE",
	id,
	updates,
});
