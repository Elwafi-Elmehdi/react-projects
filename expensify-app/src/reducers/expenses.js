// Expense Reducer

const expenseInitState = [];

export default (state = expenseInitState, action) => {
	// Never mutate the state , always return a new state
	switch (action.type) {
		case "ADD_EXPENSE":
			return [...state, action.expense];
		case "REMOVE_EXPENSE":
			return state.filter(({ id }) => id !== action.id);
		case "EDIT_EXPENSE":
			return state.map((expense) => {
				if (action.id === expense.id) {
					return {
						...expense,
						...action.updates,
					};
				} else {
					return expense;
				}
			});
		default:
			return state;
	}
};
