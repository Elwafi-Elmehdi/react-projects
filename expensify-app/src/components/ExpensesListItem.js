import React from "react";
import { connect } from "react-redux";
import { removeExpense } from "../actions/expenses";
import { Link } from "react-router-dom";

const ExpensesListItem = ({ dispatch, expense }) => (
	<div>
		<h3>{expense.title}</h3>
		<h3>{expense.createdAt}</h3>
		<h3>{expense.amount}</h3>
		<Link to={`/edit/${expense.id}`}>Edit</Link>
		<button
			onClick={() => {
				dispatch(removeExpense({ id: expense.id }));
			}}
		>
			Remove
		</button>
	</div>
);

export default connect()(ExpensesListItem);
