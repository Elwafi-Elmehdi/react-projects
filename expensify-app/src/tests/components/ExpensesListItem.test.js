import React from "react";
import ExpensesListItem from "../../components/ExpensesListItem";
import { shallow } from "enzyme";
import { expenses } from "../selectors/expenses.test";

test("Should render expenselist item with props", () => {
	const wrapper = shallow(<ExpensesListItem expense={expenses[1]} />);
	expect(wrapper).toMatchSnapshot();
});
