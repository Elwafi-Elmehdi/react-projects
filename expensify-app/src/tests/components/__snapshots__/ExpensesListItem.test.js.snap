// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render expenselist item with props 1`] = `
<div>
  <Link
    replace={false}
    to="/edit/2"
  >
    <h3>
      Car Payment bill
    </h3>
  </Link>
  <h3>
    -345600000
  </h3>
  <h3>
    3000
  </h3>
</div>
`;
